frb_rust/src/thread.rs

```diff
@@ -1,34 +1,56 @@
-pub const WORKERS_COUNT: usize = 4;
-
 #[cfg(not(wasm))]
 mod io {
-    use super::*;
-    use lazy_static::lazy_static;
+    use log::error;
     use parking_lot::Mutex;
     use threadpool::ThreadPool;
 
-    lazy_static! {
-        pub static ref THREAD_POOL: Mutex<ThreadPool> = Mutex::new(ThreadPool::with_name(
-            "frb_workerpool".into(),
-            WORKERS_COUNT
-        ));
+    pub static THREAD_POOL_HOLDER: Mutex<Option<ThreadPool>> = Mutex::new(None);
+
+    pub fn create_pool(workers_count: i32) {
+        let mut mutex_guard = THREAD_POOL_HOLDER.lock();
+
+        match *mutex_guard {
+            None => {
+                *mutex_guard = Some(ThreadPool::with_name(
+                    "frb_workerpool".into(),
+                    workers_count as usize,
+                ));
+            }
+            Some(_) => {
+                error!("Thread pool already exists (non-WASM)");
+            }
+        };
```

Comment:
Looks much better now!

Is it possible that, we somehow eliminate the whole *global variable* (THREAD_POOL_HOLDER), and put it as a field of the Handler struct? By doing so, the code style and maintainability will be higher.

Reply:
Thanks! I will be working on that In a day or two.

Reply:
You are welcome and take your time!

Reply:
IMHO it seemed best to have the threadpool or the workerpool as a field of `ThreadPoolExecutor`. I'm not sure how I should make `spawn!` macros in that case, but I'll give it a try.

Reply:
I also wanted to ask if `Mutex<ThreadPool>` or `thread_local!{WorkerPool}` is necessary. I thought that if the Rust function(async) is called from the main Dart thread, we don't need to use `Mutex` or `thread_local!` because the caller is always on the main thread. Is this approach correct?

Reply:
> I'm not sure how I should make spawn! macros work in that case, but I'll give it a try.

cc @Desdaemon - what do you think? IIRC the `spawn!` macro is introduced by your PR

> I also wanted to ask if Mutex<ThreadPool> or thread_local!{WorkerPool} is necessary

Feel free to try to remove Mutex etc! If Rust compiler is happy and there is no unsafe code, we are safe to go :)

Reply:
Since `spawn!` depends on a thread pool visible from global scope, it does make it harder to call it without a handle on the thread pool. What we could do is assume that `FLUTTER_RUST_BRIDGE_HANDLER` is in scope (i.e. defined by user or imported from bridge file) as well as being the right type, so that compile errors are emitted when the thread pool cannot be accessed from the handler. This is a breaking change, and should be communicated as such to the consumers.

Reply:
@Desdaemon Or, maybe we should add an argument to `spawn!`, i.e. explicitly pass that handler/thread-pool to it. Because global variables are known to be against the best programming styles.

Reply:
Idea of passing threadpool as an argument looks cleaner, 👌. I'm not working on it yet so design changes can take place.

Reply:
That's fine, the only reason I didn't mention passing the thread pool is it would end up looking like `spawn!(HANDLER.thread_pool, || { .. })` which seems counter-intuitive and bit too tightly coupled IMO.

Least to say, this is a shared problem when the larger ecosystem of `std::thread::spawn` not being replaceable.

Reply:
Given that macros are just converted into plain code at compile time instead of working like a function, maybe I should just find a way to access global `FLUTTER_RUST_BRIDGE_HANDLER` in `api.rs` or `bridge_generated.rs` within `spawn!`

Reply:
After diving in to the code generation mechanism, I kind of became skeptical about this idea of `threadpool per handler`.

- If `ThreadPool` becomes a field somewhere inside `SimpleHandler`, in theory user can create multiple threadpools if they make multiple handlers. In this situation, `spawn!` cannot determine which threadpool it should use(unless passing threadpool as an argument, which is not intuitive).
- If `ThreadPool` remains to be a global variable in `thread.rs` in `frb_rust` crate folder, we can assure that there is only and only one threadpool dedicated for flutter_rust_bridge function calls.

For example, if the user has multiple custom library crates each of which having `FLUTTER_RUST_BRIDGE_HANDLER` in `bridge_generated.rs`, there would be multiple threadpools because there are multiple handlers. This can be excessive, since threadpool is already a collection of multiple threads.

I am also skeptical about my approach because if the user has multiple custom library crates, it is unclear in which library should the user use the function `create_pool(N)`.

IMHO current system is brilliant and keeping it would be better. Although global variable is not the best programming style, it can still serve as a single source of data.

I just hope this 'system' can be customized, such as changing the thread count. There could be a number of ways...
- Defining FRB's own configuration file(not mandatory), located in user's project root or `./native`, where the user can customize the number of threads in the pool, log threshold level, etc.
- Environment variables.
- Reading `build.rustflags` object in `.cargo/config.toml` file.

Reply:
Hmm that's a problem.

Alternatively, we may refer to how other crates solve the problem. For example, `rayon` is a parallel computing library and has a global thread pool - similar to our case! It solves it by: https://github.com/rayon-rs/rayon/blob/master/FAQ.md#how-many-threads-will-rayon-spawn

> If you want to alter the number of threads spawned, you can set the environmental variable RAYON_NUM_THREADS to the desired number of threads or use the [ThreadPoolBuilder::build_global function](https://docs.rs/rayon/*/rayon/struct.ThreadPoolBuilder.html#method.build_global) method.

where the [build_global](https://docs.rs/rayon/1.6.1/rayon/struct.ThreadPoolBuilder.html#method.build_global) mentions it is allowed to be called exactly once